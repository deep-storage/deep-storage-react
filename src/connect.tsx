import * as React from "react";

import { DeepStorage, Path, Subscriber } from "deep-storage";

export const connect = <
  ConnectedComponentPropType extends {},
  BaseComponentPropsType extends {}
>(
  deepProps: {
    [key in keyof BaseComponentPropsType]?: DeepStorage<
      BaseComponentPropsType[key]
    >
  },
  ownProps?: {
    [key in keyof BaseComponentPropsType]?: BaseComponentPropsType[key]
  },
  additionalStorage: Array<DeepStorage<any>> = []
) => (
  BaseComponent: React.ComponentType<BaseComponentPropsType>
): React.ComponentType<ConnectedComponentPropType> => {
  const deepPropsKeys = Object.keys(deepProps) as Array<
    keyof BaseComponentPropsType
  >;

  // if no deep props specified, just return regular component
  if (deepPropsKeys.length === 0 && additionalStorage.length === 0) {
    return class extends React.Component<ConnectedComponentPropType, {}> {
      public render() {
        return <BaseComponent {...ownProps || {}} {...this.props} />;
      }
    };
  }

  // tslint:disable-next-line:max-classes-per-file
  return class extends React.Component<ConnectedComponentPropType, {}> {
    public subscriber: Subscriber = new Subscriber();

    public componentDidMount() {
      this.subscriber.onChange(() => {
        this.forceUpdate();
      });
      for (const deepPropsKey of deepPropsKeys) {
        const deepPropsValue = deepProps[deepPropsKey];
        if (deepPropsValue) {
          deepPropsValue.addSubscriber(this.subscriber);
        }
      }
      for (const storage of additionalStorage) {
        storage.addSubscriber(this.subscriber);
      }
    }
    public componentWillUnmount() {
      this.subscriber.onChange(() => {
        // noop
      });
      for (const deepPropsKey of deepPropsKeys) {
        const deepPropsValue = deepProps[deepPropsKey];
        if (deepPropsValue) {
          deepPropsValue.removeSubscriber(this.subscriber);
        }
      }
      for (const storage of additionalStorage) {
        storage.removeSubscriber(this.subscriber);
      }
    }
    public shouldComponentUpdate(
      nextProps: ConnectedComponentPropType,
      nextState: {}
    ) {
      const nextPropsAny: any = nextProps;
      for (const key in nextPropsAny) {
        if (nextPropsAny.hasOwnProperty(key)) {
          const deepPropsValue = deepProps[key as keyof BaseComponentPropsType];
          if (deepPropsValue) {
            if (deepPropsValue.state !== nextPropsAny[key]) {
              return true;
            }
          }
        }
      }
      return false;
    }
    public render() {
      const anyProps: any = this.props;
      const newProps: any = Object.assign({}, anyProps, ownProps || {});
      for (const deepPropsKey in deepProps) {
        if (deepProps.hasOwnProperty(deepPropsKey)) {
          const value = deepProps[deepPropsKey];
          if (value) {
            newProps[deepPropsKey] = value.state;
          }
        }
      }
      return <BaseComponent {...newProps} />;
    }
  };
};
