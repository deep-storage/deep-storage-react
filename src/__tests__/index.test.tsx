import { deepStorage } from "deep-storage";
import * as React from "react";
import renderer from "react-test-renderer";
import { connect, wire } from "../connect";

test("connect", async () => {
  interface IProps {
    message1: string;
    message2: string;
    className: string;
  }
  const Component: React.SFC<IProps> = props => (
    <div>
      <div className={props.className}>{props.message1}</div>
      <div className={props.className}>{props.message2}</div>
    </div>
  );
  const storage1 = deepStorage({
    message1: "message 1"
  });
  const storage2 = deepStorage({
    message2: "message 2"
  });
  const Connected = connect<{}, IProps>(
    {
      message1: storage1.deep("message1"),
      message2: storage2.deep("message2")
    },
    {
      className: "test"
    }
  )(Component);
  const component = <Connected />;
  await storage2.deep("message2").set("updated message 2");
  const rendered1 = renderer.create(component).toJSON();
  expect(rendered1).toMatchSnapshot("start");
  await storage2.deep("message2").set("updated message 3");
  const rendered2 = renderer.create(component).toJSON();
  expect(rendered2).toMatchSnapshot("update");
});
test("connect", async () => {
  interface IProps {
    message1: string;
    message2: string;
    className: string;
  }
  const Component: React.SFC<IProps> = props => (
    <div>
      <div className={props.className}>{props.message1}</div>
      <div className={props.className}>{props.message2}</div>
    </div>
  );
  const storage1 = deepStorage({
    message1: "message 1"
  });
  const storage2 = deepStorage({
    message2: "message 2"
  });
  const Connected = connect<{}, IProps>(
    {
      message1: storage1.deep("message1"),
      message2: storage2.deep("message2")
    },
    {
      className: "test"
    }
  )(Component);
  const component = <Connected />;
  await storage2.deep("message2").set("updated message 2");
  const rendered1 = renderer.create(component).toJSON();
  expect(rendered1).toMatchSnapshot("start");
  await storage2.deep("message2").set("updated message 3");
  const rendered2 = renderer.create(component).toJSON();
  expect(rendered2).toMatchSnapshot("update");
});

test("wire", async () => {
  interface IProps {
    message1: string;
    message2: string;
    className: string;
  }
  const Component: React.SFC<IProps> = props => (
    <div>
      <div className={props.className}>{props.message1}</div>
      <div className={props.className}>{props.message2}</div>
    </div>
  );
  const storage1 = deepStorage({
    message1: "message 1"
  });
  const storage2 = deepStorage({
    message2: "message 2"
  });
  const Connected = wire(
    Component,
    {
      message1: storage1.deep("message1"),
      message2: storage2.deep("message2")
    },
    {
      className: "test"
    }
  );
  const component = <Connected />;
  await storage2.deep("message2").set("updated message 2");
  const rendered1 = renderer.create(component).toJSON();
  expect(rendered1).toMatchSnapshot("start");
  await storage2.deep("message2").set("updated message 3");
  const rendered2 = renderer.create(component).toJSON();
  expect(rendered2).toMatchSnapshot("update");
});

test("wire types", async () => {
  interface IProps {
    message1: string;
    message2: string;
    className: string;
  }
  const Component: React.SFC<IProps> = props => (
    <div>
      <div className={props.className}>{props.message1}</div>
      <div className={props.className}>{props.message2}</div>
    </div>
  );
  interface IDeepState {
    message: string;
  }
  const storage = deepStorage<IDeepState>({
    message: "message"
  });
  const Connected = wire(
    Component,
    {},
    {
      message1: "one",
      message2: "two",
      className: "test"
    },
    [storage]
  );
  const component = <Connected />;
  await storage.deep("message").set("updated message");
  const rendered2 = renderer.create(component).toJSON();
  expect(rendered2).toMatchSnapshot("update");
});
