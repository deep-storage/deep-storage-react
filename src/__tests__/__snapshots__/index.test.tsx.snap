// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect: start 1`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 2
  </div>
</div>
`;

exports[`connect: start 2`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 2
  </div>
</div>
`;

exports[`connect: update 1`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 3
  </div>
</div>
`;

exports[`connect: update 2`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 3
  </div>
</div>
`;

exports[`wire: start 1`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 2
  </div>
</div>
`;

exports[`wire: update 1`] = `
<div>
  <div
    className="test"
  >
    message 1
  </div>
  <div
    className="test"
  >
    updated message 3
  </div>
</div>
`;
